contracts/eosio.bios.tonomy/src/eosio.bios.tonomy.cpp

```diff
@@ -0,0 +1,113 @@
+#include <eosio.bios.tonomy/eosio.bios.tonomy.hpp>
+
+namespace eosiobiostonomy
+{
+   // This action can only be called by inline action from the {sender} account
+   void check_sender(name sender)
+   {
+      check(eosio::get_sender() == sender, "You cannot call this eosio action directly, call from the " + sender.to_string() + " contract");
+   }
+
+   void bios::newaccount(name creator, name name, ignore<authority> owner, ignore<authority> active)
+   {
+      // require_auth(creator); // this is done implicity in apply_eosio_newaccount() which checks the "active" permission...
+      check_sender("id.tonomy"_n);
+   }
+
+   void bios::updateauth(ignore<name> account,
+                         ignore<name> permission,
+                         ignore<name> parent,
+                         ignore<authority> auth)
+   {
+      // check(has_auth({account, permission}) || has_auth({account, parent}), "not authorized by parent or permission"); // this is done implicity in apply_eosio_updateauth()...
+      check_sender("id.tonomy"_n);
+   }
+
+   void bios::deleteauth(ignore<name> account,
+                         ignore<name> permission)
+   {
+      // require_auth({account, permission}); // this is done implicity in apply_eosio_deleteauth()...
+      check_sender("id.tonomy"_n);
+   }
+
+   void bios::linkauth(ignore<name> account,
+                       ignore<name> code,
+                       ignore<name> type,
+                       ignore<name> requirement)
+   {
+      check_sender("id.tonomy"_n);
+   }
+
+   void bios::unlinkauth(ignore<name> account,
+                         ignore<name> code,
+                         ignore<name> type)
+   {
+      check_sender("id.tonomy"_n);
+   }
+
+   // TODO need to change so that other functions can only be called by tonomy logic
+
+   void bios::setabi(name account, const std::vector<char> &abi)
+   {
+      abi_hash_table table(get_self(), get_self().value);
+      auto itr = table.find(account.value);
+      if (itr == table.end())
+      {
+         table.emplace(account, [&](auto &row)
+                       {
```

Comment:
why using emplace instead of Upset ?

Reply:
We are currently using eosio v2.0.x, which only supports multi_index_table's as a way to store persistent data.
https://developers.eos.io/welcome/v2.0/smart-contract-guides/data-persistence

Newer eosio 2.1+ Supports a new data persistence mechanism called key value, which uses a new interface including the upsert function
https://github.com/EOSIO/eos/releases/tag/v2.1.0
While this feature does have several advantages, developers the upcoming fork of eosio called "Antelope" said that they will not necessarily support all features of 2.1 and will support 2.0 so best not to use any of these new features yet till we have closure on their support.
https://github.com/AntelopeIO/leap/releases/tag/v3.1.0